test/integration/volume/attach_detach_test.go

```diff
@@ -138,6 +144,183 @@ var defaultTimerConfig = attachdetach.TimerConfig{
+	server := kubeapiservertesting.StartTestServerOrDie(t, nil, []string{"--disable-admission-plugins=ServiceAccount"}, framework.SharedEtcd())
+	defer server.TearDownFn()
+
+	namespaceName := "test-volume-detach"
+	testClient, ctrl, pvCtrl, gcCtrl, informers := createAdClients(t, server, defaultSyncPeriod, defaultTimerConfig)
+	ns := framework.CreateNamespaceOrDie(testClient, namespaceName, t)
+	defer framework.DeleteNamespaceOrDie(testClient, ns, t)
+
+	// create a node
+	node := &v1.Node{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "node-sandbox",
+			Annotations: map[string]string{
+				util.ControllerManagedAttachAnnotation: "true",
+			},
+		},
+	}
+	if _, err := testClient.CoreV1().Nodes().Create(context.TODO(), node, metav1.CreateOptions{}); err != nil {
+		t.Fatalf("Failed to created node : %v", err)
+	}
+
+	// create fake pods with volumes
+	pod := fakePodWithVol(namespaceName)
+	if _, err := testClient.CoreV1().Pods(ns.Name).Create(context.TODO(), pod, metav1.CreateOptions{}); err != nil {
+		t.Errorf("Failed to create pod : %v", err)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	podInformer := informers.Core().V1().Pods().Informer()
+	// start the informer
+	informers.Start(ctx.Done())
+	informers.WaitForCacheSync(ctx.Done())
+
+	// run the controllers
+	go ctrl.Run(ctx)
+
+	// Run pvCtrl to avoid leaking goroutines started during its creation.
+	go pvCtrl.Run(ctx)
+	// Run gcCtrl to avoid leaking goroutines during its creation
+	go gcCtrl.Run(ctx)
+
+	waitToObservePods(t, podInformer, 1)
+
+	// mimic a non-graceful node shutdown
+	notReadyCOndition := v1.NodeCondition{
+		Type:   v1.NodeReady,
+		Status: v1.ConditionFalse,
+	}
+	// deep copy the node object.
+	deepCopyNode := node.DeepCopy()
+	deepCopyNode.Status.Conditions = append(deepCopyNode.Status.Conditions, notReadyCOndition)
+	taint := v1.Taint{
+		Key:    v1.TaintNodeOutOfService,
+		Effect: v1.TaintEffectNoExecute,
+	}
+	gotNode, err := testClient.CoreV1().Nodes().UpdateStatus(context.TODO(), deepCopyNode, metav1.UpdateOptions{})
+	if err != nil {
+		t.Fatalf("Failed to patch the node status : %v", err)
+	}
+	gotNode.Spec.Taints = append(gotNode.Spec.Taints, taint)
+	if _, err := testClient.CoreV1().Nodes().Update(context.TODO(), gotNode, metav1.UpdateOptions{}); err != nil {
+		t.Fatalf("Failed to taint the node : %v", err)
+	}
+
+	waitForNodeToBeNotReady(t, testClient, "node-sandbox")
+	waitForNodeToBeTainted(t, testClient, "node-sandbox", v1.TaintNodeOutOfService)
+	graceTime := int64(300)
+	err = testClient.CoreV1().Pods(namespaceName).Delete(context.TODO(), pod.Name, metav1.DeleteOptions{
+		GracePeriodSeconds: &graceTime,
+	})
+	if err != nil {
+		t.Fatalf("error in deleting pod: %v", err)
+	}
+	waitForPodDeletionTimeStampToSet(t, testClient, pod.Name, namespaceName)
+	waitForMetric(t, metrics.ForceDetachMetricCounter.WithLabelValues(metrics.ForceDetachReasonOutOfService), 1, "detach-metrics")
```

Comment:
metrics are not thread safe, if there are other tests in the same binary they will use the same global registry, please check there is no possibility of test pollution, I can't remember if the metrics registry is already initialized when the apiserver starts

Reply:
Looking into it.

Reply:
Thanks for the heads up re the metrics. 
I have seen existing tests using metrics with `gotCountofEvent>=expectedCountOfEvent` for similar reason and `waitForMetric` function does the same. 
cc @xing-yang